File: internal/analyzer/debt.go
package analyzer

import (
	"fmt"
	"math"
	"regexp"

	"github.com/codebench/codebench/internal/config"
	"github.com/codebench/codebench/internal/parser"
)

type DebtAnalyzer struct{}

var debtMarkerPattern = regexp.MustCompile(`(?i)\b(TODO|FIXME|HACK|XXX)\b`)

func (a *DebtAnalyzer) Name() config.MetricName {
	return config.MetricDebt
}

func (a *DebtAnalyzer) Analyze(files []*parser.ParsedFile, cfg *config.Config, cwd string) (*Result, error) {
	var details []Detail
	totalMarkers := 0
	totalLines := 0

	for _, file := range files {
		totalLines += file.LineCount

		for i, line := range file.Lines {
			matches := debtMarkerPattern.FindAllString(line, -1)
			if len(matches) > 0 {
				totalMarkers += len(matches)
				for _, match := range matches {
					trimmed := line
					if len(trimmed) > 100 {
						trimmed = trimmed[:100]
					}
					details = append(details, Detail{
						File:    file.RelativePath,
						Message: fmt.Sprintf("%s: %s", match, trimmed),
						Line:    i + 1,
					})
				}
			}
		}
	}

	density := 0.0
	if totalLines > 0 {
		density = float64(totalMarkers) / float64(totalLines) * 1000
	}

	maxDensity := cfg.Thresholds.MaxDebtDensity
	score := int(math.Round(math.Max(0, math.Min(100, 100-(density/maxDensity)*100))))

	// Cap details
	if len(details) > 50 {
		details = details[:50]
	}

	return &Result{
		Metric:  config.MetricDebt,
		Score:   score,
		Details: details,
		Summary: fmt.Sprintf("%d debt markers found (density: %.1f per 1K LOC)", totalMarkers, density),
	}, nil
}

File: src/analyzers/debt.ts
import type { Analyzer, AnalyzerResult, AnalyzerDetail } from './types.js';
import type { CodeBenchConfig } from '../config/index.js';
import type { ParsedFile } from '../parsers/index.js';

const DEBT_MARKERS = ['TODO', 'FIXME', 'HACK', 'XXX'];
const MARKER_PATTERN = new RegExp(`\\b(${DEBT_MARKERS.join('|')})\\b`, 'gi');

export const debtAnalyzer: Analyzer = {
  name: 'debt',

  async analyze(
    files: ParsedFile[],
    config: CodeBenchConfig,
  ): Promise<AnalyzerResult> {
    const details: AnalyzerDetail[] = [];
    let totalMarkers = 0;
    let totalLines = 0;

    for (const file of files) {
      totalLines += file.lineCount;

      for (let i = 0; i < file.lines.length; i++) {
        const line = file.lines[i];
        const matches = line.match(MARKER_PATTERN);
        if (matches) {
          totalMarkers += matches.length;
          for (const match of matches) {
            details.push({
              file: file.relativePath,
              message: `${match.toUpperCase()}: ${line.trim().substring(0, 100)}`,
              line: i + 1,
            });
          }
        }
      }
    }

    // Density = markers per 1000 LOC
    const density = totalLines > 0 ? (totalMarkers / totalLines) * 1000 : 0;
    const maxDensity = config.thresholds.max_debt_density;

    // Score: inverse of density ratio
    const score = Math.round(Math.max(0, Math.min(100, 100 - (density / maxDensity) * 100)));

    return {
      metric: 'debt',
      score,
      details: details.slice(0, 50), // cap details at 50
      summary: `${totalMarkers} debt markers found (density: ${density.toFixed(1)} per 1K LOC)`,
    };
  },
};
